src/migration.py

```diff
@@ -0,0 +1,290 @@
+    cwd = Path()
+    for row in table_rows:
+        # Next set of hierarchies, change cwd path
+        if row.level <= level:
+            if not last_dir_has_file and last_dir_row is not None:
+                yield types_.GitkeepMeta(path=cwd / GITKEEP_FILE, table_row=last_dir_row)
+            while row.level <= level:
+                level -= 1
+                cwd = cwd.parent
+
+        # if row is directory, move cwd
+        if not row.navlink.link:
+            last_dir_has_file = False
+            last_dir_row = row
+            cwd = cwd / row.path
+            level = row.level
+        else:
+            last_dir_has_file = True
+            file_name = _calculate_file_name(cwd, row.path)
+            yield types_.DocumentMeta(
+                path=cwd / f"{file_name}.md", link=row.navlink.link, table_row=row
+            )
+
+    if not last_dir_has_file and last_dir_row:
+        yield types_.GitkeepMeta(path=cwd / GITKEEP_FILE, table_row=last_dir_row)
+
+
+def _index_file_from_content(content: str):
+    """Get index file document metadata.
+
+    Args:
+        content: Index file content.
+
+    Returns:
+        Index file document metadata.
+    """
+    return types_.IndexDocumentMeta(path=Path("index.md"), content=content)
+
+
+def get_docs_metadata(
+    table_rows: typing.Iterable[types_.TableRow], index_content: str
+) -> typing.Iterable[types_.MigrationFileMeta]:
+    """Get metadata for documents to be migrated.
+
+    Args:
+        table_rows: Table rows from the index table.
+        index_content: Index content from index page.
+
+    Returns:
+        Metadata of files to be migrated.
+    """
+    index_doc = _index_file_from_content(content=index_content)
+    table_docs = _extract_docs_from_table_rows(table_rows=table_rows)
+    return itertools.chain([index_doc], table_docs)
+
+
+def run(
+    documents: typing.Iterable[types_.MigrationFileMeta], discourse: Discourse, docs_path: Path
+) -> typing.Iterable[types_.MigrationReport]:
+    """Write document content to docs_path.
+
+    Args:
+        documents: metadata about a file to be migrated to local docs directory.
+        discourse: Client to the documentation server.
+        docs_path: The path to the docs directory containing all the documentation.
+
+    Returns:
+        Migration result reports containing action result and failure reason if any.
+    """
+    return [
+        _run_one(file_meta=document, discourse=discourse, docs_path=docs_path)
+        for document in documents
+    ]
+
+
+def assert_migration_success(migration_results: typing.Iterable[types_.MigrationReport]) -> None:
+    """Assert all documents have been successfully migrated.
+
+    Args:
+        migration_results: Migration results from server to local.
+
+    Returns:
+        None if success, raises MigrationError otherwise.
+    """
+    if [result for result in migration_results if result.result is types_.ActionResult.FAIL]:
```

Comment:
`result.result` is a bit confusing, would it be worth renaming? Maybe `migration_reports` and `report.result`? That would align with the current `ActionReport`. Do you think that data structure could be used here?